mirrord/cli/src/wizard.rs

```diff
@@ -0,0 +1,676 @@
+//! # mirrord Wizard (aka onboarding Wizard)
+//!
+//! The `wizard` module contains everything needed to run the `mirrord wizard` CLI command, with the
+//! exception of the frontend code. This lives in a different top-level directory called
+//! `wizard-frontend`. More documentation specific to the frontend exists in
+//! `wizard-frontend/README.md`. The `wizard` module is gated by the `wizard` feature.
+//!
+//! **The Wizard operates as follows**:
+//!
+//! 1. Build the TypeScript frontend and compress `wizard-frontend/dist` into a Gzipped file
+//! 2. Use the `include_bytes!` macro to make [`COMPRESSED_FRONTEND`] when compiled with the
+//!    `wizard` feature
+//! 3. When the command is run, the file is unzipped and the frontend served on localhost with axum.
+//! 4. The wizard opens in the browser and the user can create and download a config file
+//!
+//! ### Working on the Backend
+//!
+//! Note: To compile the backend, you need to enable the `wizard` feature.
+//!
+//! The build script `build.rs` installs node modules, builds the frontend, and compresses it so it
+//! can be accessed by [`COMPRESSED_FRONTEND`].
+//!
+//! Node modules are installed into the `wizard-frontend` top level dir (and are `.gitignore`d)
+
+use std::{
+    io::Cursor,
+    net::{Ipv4Addr, SocketAddr},
+    ops::Not,
+    sync::Arc,
+};
+
+use axum::{
+    Router,
+    extract::{Path, Query, State},
+    http::header,
+    routing::get,
+};
+use flate2::read::GzDecoder;
+use futures::{StreamExt, TryStreamExt};
+use itertools::Itertools;
+use k8s_openapi::api::{
+    apps::v1::{Deployment, ReplicaSet, StatefulSet},
+    batch::v1::{CronJob, Job},
+    core::v1::{Namespace, Pod, Service},
+};
+use kube::{Client, client::ClientBuilder, runtime::reflector::Lookup};
+use mirrord_analytics::{AnalyticsReporter, ExecutionKind};
+use mirrord_config::target::TargetType;
+use mirrord_kube::{
+    api::kubernetes::{create_kube_config, rollout::Rollout, seeker::KubeResourceSeeker},
+    error::KubeApiError,
+};
+use mirrord_progress::Progress;
+use serde::{Deserialize, Serialize};
+use tar::Archive;
+use tempfile::TempDir;
+use tokio::sync::Mutex;
+use tower::ServiceBuilder;
+use tower_http::{
+    services::{ServeDir, ServeFile},
+    set_header::SetResponseHeaderLayer,
+    trace::TraceLayer,
+};
+
+use crate::{
+    config::WizardArgs,
+    error::{CliError, CliResult},
+    user_data::UserData,
+};
+
+/// The frontend `dist` dir, compressed (as bytes). The CI runs this step automatically, so you only
+/// need to do it manually if making changes to the wizard.
+const COMPRESSED_FRONTEND: &[u8] =
+    include_bytes!(concat!(env!("OUT_DIR"), "/wizard-frontend.tar.gz"));
+
+/// Used to store State for endpoint handling
+struct BackendState {
+    // for updating is-returning in `UserData`
+    user_data: UserData,
```

Comment:
Probably won't matter much in our case. I wonder if you have considered loading `UserData` from file each time you need `UserData` in a handler. Then, we can get rid of the mutex.